lib/validation_deso_txn.go

```diff
@@ -0,0 +1,166 @@
+	// Validate public key
+	if err := IsByteArrayValidPublicKey(txn.PublicKey); err != nil {
+		return errors.Wrapf(err, "ValidateDeSoTxnPublicKey: Problem with public key")
+	}
+	return nil
+}
+
+// ValidateDeSoTxnFormatBalanceModel validates that the transaction is correctly formatted according to the balance model.
+func ValidateDeSoTxnFormatBalanceModel(txn *MsgDeSoTxn, blockHeight uint64, globalParams *GlobalParamsEntry) error {
+	var err error
+
+	// Validate transaction version
+	if txn.TxnVersion == DeSoTxnVersion0 {
+		return fmt.Errorf("ValidateDeSoTxnFormatBalanceModel: DeSoTxnVersion0 is outdated in balance model")
+	}
+
+	if txn.TxnNonce == nil {
+		return errors.Wrapf(TxErrorNoNonceAfterBalanceModelBlockHeight, "ValidateDeSoTxnFormatBalanceModel: Transaction "+
+			"does not have a nonce.")
+	}
+	if txn.TxnNonce.ExpirationBlockHeight < blockHeight {
+		return errors.Wrapf(TxErrorNonceExpired, "ValidateDeSoTxnFormatBalanceModel: Transaction nonce has expired")
+	}
+	if globalParams.MaxNonceExpirationBlockHeightOffset != 0 &&
+		txn.TxnNonce.ExpirationBlockHeight > blockHeight+globalParams.MaxNonceExpirationBlockHeightOffset {
+		return errors.Wrapf(TxErrorNonceExpirationBlockHeightOffsetExceeded, "ValidateDeSoTxnFormatBalanceModel: Transaction "+
+			"nonce expiration block height offset exceeded")
+	}
+
+	// Verify inputs/outputs.
+	if len(txn.TxInputs) != 0 {
+		return errors.Wrapf(RuleErrorBalanceModelDoesNotUseUTXOInputs, "ValidateDeSoTxnFormatBalanceModel: Balance model "+
+			"transactions should not have any inputs")
+	}
+
+	// Loop through the outputs and do a few sanity checks.
+	var totalOutNanos uint64
+	for _, txout := range txn.TxOutputs {
+		// Check that each output's amount is not bigger than the max as a
+		// sanity check.
+		if txout.AmountNanos > MaxNanos {
+			return errors.Wrapf(RuleErrorOutputExceedsMax, "ValidateDeSoTxnFormatBalanceModel: Output amount %d "+
+				"exceeds max %d", txout.AmountNanos, MaxNanos)
+		}
+		// Check that this output doesn't overflow the total as a sanity
+		// check. This is frankly impossible since our maximum limit is
+		// not close to the max size of a uint64 but check it nevertheless.
+		if totalOutNanos, err = SafeUint64().Add(totalOutNanos, txout.AmountNanos); err != nil {
+			return errors.Wrapf(RuleErrorOutputOverflowsTotal, "ValidateDeSoTxnFormatBalanceModel: Output amount %d "+
+				"overflows total %d", txout.AmountNanos, totalOutNanos)
+		}
+		// Check that the total isn't bigger than the max supply.
+		if totalOutNanos > MaxNanos {
+			return errors.Wrapf(RuleErrorTotalOutputExceedsMax, "ValidateDeSoTxnFormatBalanceModel: Total output "+
+				"amount %d exceeds max %d", totalOutNanos, MaxNanos)
+		}
+	}
+	return nil
+}
+
+// ValidateDeSoTxnMinimalNetworkFee validates that the transaction is above the network's minimal fee.
+func ValidateDeSoTxnMinimalNetworkFee(txn *MsgDeSoTxn, globalParams *GlobalParamsEntry) error {
```

Comment:
With PoS we're not going to have a minimal network fee anymore. Instead, the fee can be ZERO. This will probably be the case for some time (until blocks start getting filled). So I would just remove this check entirely.

Instead, node mempools will be responsible for putting higher-fee txns at the front of the queue and dropping spam as needed.

Reply:
This minimum fee is also for backwards-compatibility. It's not meant to invalidate any meaningful transactions, rather set a really small lowerbound for the possible transaction fees. Many things are easier when the fee rate can't be 0, and having no minimal fee forces us to gate against this edge-case elsewhere. For instance Fee buckets won't work without non-zero starting fee. If we set the bucket starting fee to 1 nano, then it'll never grow, so it needs to be at least 1/growthRate, otherwise the growth will be <1 which may not round up.

The minimum fee in question would be like 1000 nanos, really small and noone would ever hit it. I think it's easiest to keep this one as it is.

Reply:
To simplify the exponential fee bucketing math, we decided some time back to have a non-zero minimal network fee. In practice, we'll set the minimum fee to 1 nano so that the fee bucketing doesn't need any addition edge case handling to work with a near-zero starting fee. Given we can have a 1 nano min fee, I don't believe a zero fee provides much more value.

Marking this as resolved based on the above.